### ISSUE TITLE
[CSL-97] MPC overhead: investigate & optimize

### ISSUE THREAD

**Issue by jagajaga** [input-output-hk/cardano-sl#481]
<blockquote>@georgeee</blockquote>

During testing under CSL-88 we encountered some problems appeared to be related to MPC message relaying.

More specifically, even with no transactions sent, nodes experienced 100% CPU load and blocks didn't get properly processed as a consequence (and hence nothing worked).

After we switched off MPC relaying (and made all nodes sent MPC messages to each node directly) CPU load became less than 5%.

In general, for MPC each node forms a message and send it as broadcast. It's `N` messages. Consider `N=60`. Given we use Kademlia DHT for P2P, we have no more than `~10-12` neighbors. This way each node sends & receives `~600` messages average. And it's still `60` for MPC relaying off scenario. How could it get that `60` messages are Ok and `600` messages break everything. Where is bottleneck.

Several hypothesizes:

1. Actual amount of messages is much bigger than `600`
To be checked by `grep '...' | wc -l` on logs
2. Time-warp
     1. Problems with synchronization (we have lot of threads executed, also conduits queues and other stuff)
     2. Ineffective/buggy message processing/deserialization
     Note that time-warp would be self examined within TW-38
3. Acid-state
Each server request, even those which doesn't update anything, requires disk operations
Following could be done:
* launch remote benchs with disk state turned off (they are launched with `--rebuild-db` now, which is pretty the same effect)
* contact to @neongreen for his profiling results (and reproduce this results locally, maybe remotely)

**Comment by jagajaga** [comment]
<blockquote>@neongreen</blockquote>

Regarding acid-state, we can reduce the disk load by... wait for it... yes, more acid-state hacks! Specifically, I can add a way to mark certain updates as “in-memory only” (and in the `SafeCopy` instance for `Storage` we would have to exclude memory-only fields from being written to disk, but that's easy).

**Comment by jagajaga** [comment]
<blockquote>@neongreen</blockquote>

As for profiling results – I have the code in a slightly hacky branch. I can push it but then it'd be nice if someone could work with me on integrating the -changes I had to make- into develop. Specifically:

* A way to specify how much the node would run (if we kill it by throwing an exception, it doesn't save the profiling results for some reason, so we need to have a command-line flag to replace `sleepForever` in `runNode` with a wait).

* A way to reliably run the node with disk state turned off (the last time I tried, `--rebuild-db` didn't work, but maybe now it does).

* Easy tx stats collection (so that we would know when the profiling is useless because we're actually getting only 200tps and not 600tps, or something like that). Again, probably already done. In my branch I did it by outputting transaction counts in each generated block – not the best approach :)

* I also disabled `timestampsTxSender.json` log writing in tx-gen just in case (I wasn't sure how much it was writing and was afraid it was “a lot”). If it doesn't write much and doesn't hurt the tx-gen performance or introduce delays, then this isn't necessary.

**Comment by jagajaga** [comment]
<blockquote>@gromak</blockquote>

>  A way to reliably run the node with disk state turned off (the last time I tried, --rebuild-db didn't work, but maybe now it does).

`--rebuild-db` rebuilds db, it doesn't make node use memory state.

**Comment by jagajaga** [comment]
<blockquote>@georgeee</blockquote>

Sure, I meant that we don't need persistence at this point

**Comment by jagajaga** [comment]
<blockquote>@gromak</blockquote>

Today I checked that number of messages we sent is reasonable. I have found several problems:
1. Bug in `defaultSendToNeighbors` because of which messages could be sent twice. It was unlikely to happen with big number of nodes, but this bug is dangerous anyway. I have fixed it.
2. We didn't add MPC data to storage before sending it to others. Because of it the following situations happened often: we send our commitment/opening to others, they accept it, resend it to everyone including us, we accept it (because it's not in local storage) and resend it again. I don't think it could lead to big performance issues, but it's ridiculous. I have changed it.
3. When we receive message and accept it, we send it back to the node who sent it to us. It's an obvious problem which shouldn't give much overhead, so I didn't spend time on fixing it, but wrote about it here, because I wanted to mention all overhead we have. Btw, if number of neighbors is about 10, this improvement can give us 10% performance benefit, which is not negligible.

In general number of messages seems reasonable. For example, I launched 8 nodes locally, disabled `sscTransmitter` and I see that during 0-th epoch number of `DSOpenings` and `DSShares` received is 56. It's exactly what I expect (each neighbor sends us messages from everyone, including us, hence 7 * 8 = 56). Number of commitments is lower, because some nodes start before others and when send commitments before other nodes start.

What I also need to check regarding number of messages:
1. Enable `sscTransmitter` and check it.
2. Situations where number of nodes is bigger than number of neighbors.

Other problems which I realized today:
1. When we switch to strategy where we resend messages when we accept them, we effectively disabled batching, because messages are created with size 1 and then are relayed as is. So each message has size 1.
2. I think CSL-120 can really help us, but I am not 100% sure. I am going to do it and we'll see whether it is helpful.

**Comment by jagajaga** [comment]
<blockquote>@gromak</blockquote>

> Enable sscTransmitter and check it.

`sscTransmitter` also doesn't do anything insane. It really sends all data in one message. It doesn't send data which is already in blockchain.
One thought which came into my mind regarding transmitter: probably it makes sense to randomize time when transmitter starts for the same reason as CSL-120. On the other hand, in real deployment nodes start at different time (with some delays) and also it takes different amount of time to transmit messages. However, these differences should be relatively small, that's why I probably will do this change.

And more one thought: `Commitment` and `(HashMap PublicKey Share)` can be big when number of stakeholders is not too small. So we shouldn't send them every time, it makes more sense to send only meta info (e. g. `I have commitment given by this public key for this epoch`) and send actual data only on demand. Just like we do it with blocks, for example. It's not necessary for `Opening`s, because they are small.

**Comment by jagajaga** [comment]
<blockquote>@gromak</blockquote>

@neongreen sorry that it took so much time to react properly, but I think the following (regarding your hacky branch):
1.
> A way to specify how much the node would run (if we kill it by throwing an exception, it doesn't save the profiling results for some reason, so we need to have a command-line flag to replace sleepForever in runNode with a wait).
It may be useful, I agree, I will create issue (if it's easy to port your code to current version, just do it).
2. 
> A way to reliably run the node with disk state turned off (the last time I tried, --rebuild-db didn't work, but maybe now it does).
As I said above, `--rebuild-db` is irrelevant, but this thing is also useful. `Launcher` already supports it, we just need a command line option. So again, I will create an issue.
3, 4 are about our benchmarks, they are currently WIP, so I don't know.

**Comment by jagajaga** [comment]
<blockquote>@gromak</blockquote>

> Each server request, even those which doesn't update anything, requires disk operations
This problem, even if exists, will disappear after CSL-24 is finished, so I don't want to spend time on it.

**Comment by jagajaga** [comment]
<blockquote>@gromak</blockquote>

> sending an object w. prior request == ban!

This is what @kushti suggested in #paper-pos, btw.

**Comment by jagajaga** [comment]
<blockquote>@gromak</blockquote>

Btw, issues are CSL-141 and CSL-142.

**Comment by jagajaga** [comment]
<blockquote>@gromak</blockquote>

My opinion regarding this issue:
1. We must ensure that when we receive data which we are not interested in, we ignore it very quickly. I find it important, because in real network someone can send us such data quite often. We can also introduce some mechanism to ban nodes, but it's quite hard and even if node doesn't misbehave it can send us useless data quite often in some situations. I think it will be solved by CSL-24. At least we won't use disk on each request (which is ridiculous, to be honest). Other reasons which can make it slow: inefficient encoding/decoding, inefficient networking layer, too big messages (we shouldn't send all shares for example, see above).
2. We must rework protocol regarding MPC. I propose to start with Bitcoin-like approach, because it already exists, it works and @kushti recommended INV messages. I hope that together with other improvements (previous point, efficient network layer, efficient encoding/decoding) it will eliminate all performance problems which we have now (i. e. MPC overhead with 100 nodes). Protocol is as follows.
2.1. We have `SscInv` message which consists of identifiers of data which node stores locally. Identifier of data consists of type of data (i. e. `Commitment`, `Opening`, `Shares`, `VssCertificate`), `PublicKey` of party which issued this data, `EpochIndex` for which this data is actual (except `VssCertificate`). Type of data and epoch index are the same for all items in `SscInv` message (e. g. we can't have openings and commitments at the same time), only `PublicKey`s are different.
2.2. When node wants to announce her data, she sends `SscInv` consisting of one item. When node receives `SscInv` from another node, she figures out which items she is interested in (it should be cheap, see 1.) and requests them (see 2.5). When node adds some data to storage, she sends corresponding `SscInv` to her peers. It's basically what we do now and it's described [here|https://en.bitcoin.it/wiki/Network#Standard_relaying].
2.3. As you can see from 2.2, there is no way how `SscInv` with more than 1 item can appear. I wrote about it above. From Bitcoin documentation I don't understand whether in Bitcoin `inv` message with more than 1 items can appear. Maybe it can be generated by client who sends many transactions, for example. Dunno. Probably it can be optimized by sending no more than 1 message per second (for example). Then if we receive a lot of data in one second, we will propagate it in a single message. It's close to `transmitter` worker which we have now. But it's upgraded version of such worker, because it sends only recent data, not whole local set. Or maybe whole local set.
This is a hard question, I don't have a clear understanding, it requires experiments. For this reason, I propose to postpone this optimization (which may be not optimization at all) and do more clear things before. After we do main part, we will run experiments and they will show whether this «optmization» makes sense.
2.4. When node joins network, she requests `SscInv` message from her peers. *Questionale*: is it needed? We can just wait until others send us their data.
2.5. Of course nodes have endpoints using which other nodes can request ssc data from `SscInv` message. For `Commitment` and `Shares` we probably should return only 1 message per 1 request. If number of stakeholders is big, then even 1 message may be very big (and we'll need to send it in chunks), but I hope delegation scheme will solve it. For openings we can support batch requests, but they are not really needed unless we do optimization from 2.3 or do 2.4.
3. Do not send data to node which sent it to us. If it's not very hard to do, let's do it. It won't give us significant benefit, but even 5-10% is good enough.
4. It's questionable whether it makes sense to use `SscInv` for openings, because they are really small, but I think yes, because we will receive known openings much more often and in such cases we don't need actual opening.

**Comment by jagajaga** [comment]
<blockquote>@georgeee</blockquote>

> 1. We must ensure that when we receive data which we are not interested in, we ignore it very quickly.
Time for enable LRU cache for some non-broadcast messages?
Given SscInv in 99% cases would contain only a single item, that should help.

Regarding all this SscInv scheme -- ok.

**Comment by jagajaga** [comment]
<blockquote>@georgeee</blockquote>

Also, from latest bench: on 20-node setup nodes experience:
1. 0th and 1th slots (any epoch) -- 40-70% CPU // ~40sec
2. 12th slot -- no load (< 5% CPU)
3. 24th slot -- 15-40 % CPU // ~15 sec, starting from ~4th sec of slot

**Comment by jagajaga** [comment]
<blockquote>@georgeee</blockquote>

Also, as I understand `SscInv` as described above might lead to situation, when all neighbors send you same item, for all incoming `SscInv` you see that data is unknown to you, you request it from all 10 nodes and receive 10 big messages. It's well likely to occur in practice. We can modify code in such a way that single resource is requested from single provider only (first sent us `SscInv`)

**Comment by jagajaga** [comment]
<blockquote>@georgeee</blockquote>

However it's unsafe to rely only on first sender, so we should ask neighbors in turn with say 500ms intervals. But this is certainly for "later improvements"

**Comment by jagajaga** [comment]
<blockquote>@gromak</blockquote>

> Time for enable LRU cache for some non-broadcast messages?
I am not sure about it. If we store local MPC data in memory, then I don't see significant difference. It's only a lookup in map which is stored in memory (in both cases). Btw, how does LRU cache work? Does it cache raw data or decoded data? If it stores raw data, then it may be beneficial, but may be the same.

Apart from that, I am afraid that cache is bad in this case. For example: we receive commitment from `pk`, put it into cache, put it into local data, everything is great, then someone creates block with commitment, we receive it, remove this commitment from local data, then rollback happens and commitment is not in blockchain anymore, then we receive commitment from `pk` again and ignore it because of cache, but we shouldn't ignore it.

> 3. 24th slot – 15-40 % CPU // ~15 sec, starting from ~4th sec of slot

It should become close to 0 in last version where I merged CSL-25.

> However it's unsafe to rely only on first sender, so we should ask neighbors in turn with say 500ms intervals.
Yes, good optimization.

**Comment by jagajaga** [comment]
<blockquote>@georgeee</blockquote>

1. Well, now it works on top of raw data but after data gets decoded (there were reasons to make it behave in such way). But decoding is believed to be a cheap operation.

> If we store local MPC data in memory, then I don't see significant difference.
No difference at all, the only is that we don't need to implement this ignore message logic in one more place (when we already have it in communication level)

> Apart from that, I am afraid that cache is bad in this case. For example: we receive commitment from pk , put it into cache, put it into local data, everything is great, then someone creates block with commitment, we receive it, remove this commitment from local data, then rollback happens and commitment is not in blockchain anymore, then we receive commitment from pk again and ignore it because of cache, but we shouldn't ignore it.
Looks like we need to put commitment back to local data on rollback.

**Comment by jagajaga** [comment]
<blockquote>@gromak</blockquote>

> No difference at all, the only is that we don't need to implement this ignore message logic in one more place (when we already have it in communication level)

Lol, I would say vice versa :)
We need to have this logic on local data level anyway. When we receive commitment which passes cache, it doesn't mean that this commitment is not in blocks. For example, it may happen that we receive block with commitment and then commitment itself. It will pass cache, but shouldn't be added.

> Looks like we need to put commitment back to local data on rollback.
Well, apparently we can do it. It can be not so easy because of our super abstractions though.

**Comment by jagajaga** [comment]
<blockquote>@gromak</blockquote>

Protocol used for MPC communication has been crucially changed according to what I described above. Now we need to test that it really solved our problems when we have 100 nodes. CSL-24 has been done too.

There are some other ideas in comments, but they require more investigation and experiments. We can leave this issue open, but probably it makes sense to close it and create another issue for further experiments/improvements.

Current state is `To Verify`, because only benchmarks will show whether problem has been solved.
